### ISSUE TITLE
Mixture Density Network quickly gets to nan loss

### ISSUE THREAD

**Issue by sergeyf** [fchollet/keras#2586]
Hello,

I've got the latest pulls of keras and theano.

I am trying to reproduce the code from [this blog post](http://cbonnett.github.io/MDN.html).

What I have so far is in a gist [here](https://gist.github.com/sergeyf/cf20b2759a7d38035f30384769bed9df).

The issue is that after a few batch updates (on the order of 10), the loss becomes nan.

Here is a repaste of the most likely culprit: the negative log-likelihood loss function:

```
def neg_beta_mixture_likelihood(true, parameters):
    m = K.shape(parameters)[-1] // 3

    alphas = parameters[:, 0 * m: 1 * m] # nbatch by m
    betas = parameters[:, 1 * m: 2 * m] # nbatch by m 
    pis = parameters[:, 2 * m: 3 * m] # nbatch by m

    # true is nbatch by 1
    # true_repeated is nbatch by m
    true_repeated = K.repeat_elements(true, m, axis=-1)

    d1 = (alphas - 1.0) * K.log(true_repeated)
    d2 = (betas - 1.0) * K.log(1.0 - true_repeated)
    f1 = d1 + d2
    f2 = gammaln(alphas)      
    f3 = gammaln(betas)      
    f4 = gammaln(alphas + betas) 
    exponent = f1 + f4 - f2 - f3
    
    # log sum exp trick to prevent numerical issues
    max_exponent = K.max(exponent, axis=-1, keepdims=True)
    max_exponent_repeated = K.repeat_elements(max_exponent, m, axis=-1)
    likelihood = pis * K.exp( exponent - max_exponent_repeated  )  
    
    return K.mean( -(K.log(K.sum(likelihood,axis=-1)) + max_exponent), axis=-1 )
```

I've tried to alter various parts of this function, and can't spot any specific issues.

I've added the log-sum-exp trick to prevent underflow in there, and when I check out values of each intermediate variable within the loss function during any of the batches before the nan loss, they all seem very well behaved.

Any ideas?

Thanks for your help, and the fabulous package.

**Comment by braingineer** [comment]
Are you doing any sort of grad clipping or grad norm restraints? 

my optimizers typically get instantiated something like:

```python
        optimizer = Adam(self.igor.LR, clipnorm=self.igor.max_grad_norm, 
                                       clipvalue=self.igor.grad_clip_threshold)
```

I find that numbers in the mid single digits usually works decently.  

One question, though, is where do NaNs pop in at, though? Well, it turns out it usually only happens when infs are interacting ([stack overflow question on it](http://stackoverflow.com/questions/25506281/what-are-all-the-possible-calculations-that-could-cause-a-nan-in-python)).    And how do you get infs?  Well, a couple ways:

```python
import numpy as np
np.log([0]) # returns -inf
1/np.array([0.]) #returns inf
np.exp(1e100) # returns info
```

Aside from the overflow one (the exp, which you seem to be already thinking about), you should think about how you can clip your values in order to avoid zeros.  for instance, assuming `true_repeated` is on a support set of [0,1]:

```python
np.clip(true_repeated, 1e-8, 1.-1e-8)
```

Clipping where you can reasonably clip to avoid 0s would work well, I think.

Finally, if you're using theano, they have a nice page on this: http://deeplearning.net/software/theano/tutorial/nan_tutorial.html

**Comment by sergeyf** [comment]
Thanks for the great tips @braingineer !

Clipping `true_repeated` solved the problem.

**Comment by raddy** [comment]
Actually, I took a look at this as MDNs are pretty interesting to me and definitely run into NaNs pretty easily. (Which has been my experience trying to train MDNs always as you've got I'd expect super ugly gradients)

**Comment by sergeyf** [comment]
The gist has code that doesn't become NaNs now, but it still doesn't solve my toy problem, unfortunately.

**Comment by raddy** [comment]
Yeah, I'm looking at the updated gist actually-- I think it's a an interesting toy problem. It's actually very very similar to what I'm working on. 

And re NaNs, sometimes I get them, sometimes I don't. I'll keep playing with it to see where they're coming from.

**Comment by braingineer** [comment]
have you run it with nanguard mode?  I find that usually helps me find where the damn nans are coming from.
